test/recipes/70-test_tls13downgrade.t

```diff
@@ -23,72 +23,119 @@ plan skip_all => "$test_name needs the dynamic engine feature enabled"
+    # Run tests with TLS
+    run_tests(0);
+}
 
 SKIP: {
-    skip "TLSv1.1 disabled", 2 if disabled("tls1_1");
-    #Test 5: A client side protocol "hole" should not be detected as a downgrade
+    skip "DTLS 1.2 or 1.3 is disabled", $testcount if disabled("dtls1_3")
+                                                      || disabled("dtls1_2");
+    skip "DTLSProxy does not work on Windows", $testcount if $^O =~ /^(MSWin32)$/;
+    run_tests(1);
+}
+
+sub run_tests
+{
+    my $run_test_as_dtls = shift;
+    my $proto1_1 = $run_test_as_dtls == 1 ? "DTLSv1" : "TLSv1.1";
+    my $proto1_2 = $run_test_as_dtls == 1 ? "DTLSv1.2" : "TLSv1.2";
+    my $proto1_3 = $run_test_as_dtls == 1 ? "DTLSv1.3" : "TLSv1.3";
+
+    my $proxy;
+    if ($run_test_as_dtls == 1) {
+        $proxy = TLSProxy::Proxy->new_dtls(
+            undef,
+            cmdstr(app([ "openssl" ]), display => 1),
+            srctop_file("apps", "server.pem"),
+            (!$ENV{HARNESS_ACTIVE} || $ENV{HARNESS_VERBOSE})
+        );
+    }
+    else {
+        $proxy = TLSProxy::Proxy->new(
+            undef,
+            cmdstr(app([ "openssl" ]), display => 1),
+            srctop_file("apps", "server.pem"),
+            (!$ENV{HARNESS_ACTIVE} || $ENV{HARNESS_VERBOSE})
+        );
+    }
+
+    #Test 1: Downgrade from (D)TLSv1.3 to (D)TLSv1.2
+    $proxy->clear();
+    $proxy->filter(\&downgrade_filter);
+    $testtype = DOWNGRADE_TO_TLS_1_2;
+    skip "Unable to start up Proxy for tests", $testcount if !$proxy->start() &&
+                                                             !TLSProxy::Message->fail();
+    ok(TLSProxy::Message->fail(), "Downgrade ".$proto1_3." to ".$proto1_2);
+
+    #Test 2: Downgrade from (D)TLSv1.3 to TLSv1.1/DTLSv1
     $proxy->clear();
-    $proxy->filter(undef);
-    $proxy->clientflags("-no_tls1_2");
-    $proxy->ciphers("AES128-SHA:\@SECLEVEL=0");
+    $testtype = DOWNGRADE_TO_TLS_1_1;
     $proxy->start();
-    ok(TLSProxy::Message->success(), "TLSv1.2 client-side protocol hole");
+    ok(TLSProxy::Message->fail(), "Downgrade ".$proto1_3." to ".$proto1_1);
 
-    #Test 6: A server side protocol "hole" should not be detected as a downgrade
+    #Test 3: Downgrade from (D)TLSv1.2 to TLSv1.1/DTLSv1
     $proxy->clear();
-    $proxy->filter(undef);
-    $proxy->serverflags("-no_tls1_2");
+    $proxy->clientflags("-max_protocol ".$proto1_2);
+    $proxy->serverflags("-max_protocol ".$proto1_2);
     $proxy->start();
-    ok(TLSProxy::Message->success(), "TLSv1.2 server-side protocol hole");
+    ok(TLSProxy::Message->fail(), "Downgrade ".$proto1_2." to ".$proto1_1);
+
+
+    SKIP: {
+        # TODO(DTLSv1.3):
+        # The handshake is terminated with
+        #    inappropriate fallback:ssl/statem/statem_srvr.c:1822
+        # Need to determine if this is correct behaviour.
+        skip "Test does not work with second ClientHello after a HelloVerifyRequest",
+             1 if $run_test_as_dtls == 1;
```

Comment:
This is because the downgrade_filter does not filter out the SCSV in the second ClientHello after the HVR. It needs to be tweaked.